llama-index-integrations/llms/llama-index-llms-lmstudio/llama_index/llms/lmstudio/base.py

```diff
@@ -0,0 +1,245 @@
+import logging
+import json
+import httpx
+from httpx import Timeout
+from typing import Any, Dict, Sequence, Tuple
+from llama_index.core.bridge.pydantic import Field
+from llama_index.core.llms.callbacks import llm_chat_callback, llm_completion_callback
+from llama_index.core.llms.custom import CustomLLM
+
+from llama_index.core.base.llms.generic_utils import (
+    stream_chat_to_completion_decorator,
+    chat_to_completion_decorator,
+    achat_to_completion_decorator,
+    astream_chat_to_completion_decorator,
+)
+from llama_index.core.constants import (
+    DEFAULT_TEMPERATURE,
+    DEFAULT_CONTEXT_WINDOW,
+    DEFAULT_NUM_OUTPUTS,
+)
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+    ChatResponse,
+    ChatResponseGen,
+    CompletionResponse,
+    CompletionResponseAsyncGen,
+    CompletionResponseGen,
+    LLMMetadata,
+    MessageRole,
+)
+
+
+def get_additional_kwargs(
+    response: Dict[str, Any], exclude: Tuple[str, ...]
+) -> Dict[str, Any]:
+    return {k: v for k, v in response.items() if k not in exclude}
+
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_REQUEST_TIMEOUT = 30.0
+
+
+class LMStudio(CustomLLM):
+    base_url: str = Field(
+        default="http://localhost:1234/v1",
+        description="Base url the model is hosted under.",
+    )
+
+    context_window: int = Field(
+        default=DEFAULT_CONTEXT_WINDOW,
+        description="The maximum number of context tokens for the model.",
+        gt=0,
+    )
+
+    model_name: str = Field(description="The model to use.")
+
+    request_timeout: float = Field(
+        default=DEFAULT_REQUEST_TIMEOUT,
+        description="The timeout for making http request in seconds to LM Studio API server.",
+    )
+    num_output: int = Field(
+        default=DEFAULT_NUM_OUTPUTS,
+        description=LLMMetadata.__fields__["num_output"].field_info.description,
+    )
+
+    is_chat_model: bool = Field(
+        default=True,
+        description=(
+            "LM Studio API supports chat."
+            + LLMMetadata.__fields__["is_chat_model"].field_info.description
+        ),
+    )
+
+    temperature: float = Field(
+        default=DEFAULT_TEMPERATURE,
+        description=("The temperature to use for sampling."),
+        gte=0.0,
+        lte=1.0,
+    )
+
+    timeout: float = Field(
+        default=120, description=("The timeout to use in seconds."), gte=0
+    )
+
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description=("Additional kwargs to pass to the model.")
+    )
+
+    def __create_payload_from_messages(
```

Comment:
nit: lets just use single under-score for private methods

Reply:
👍 Fixed the typo